### ISSUE TITLE
about Exception: Number of user feature rows does not equal the number of users

### ISSUE THREAD

**Issue by kyowill** [lyst/lightfm#196]
code like this:
predict=model.predict(1,np.arange(2),user_features=user_features_concat[1],item_features=item_features_concat)

user_features_concat[1] is <1x45023 sparse matrix of type '<class 'numpy.float32'>'
  in lightfm.py, i checked the source code "n_users = user_ids.max() + 1" 
  as a result "n_users > user_features.shape[0]:"
Did I input wrong parameter?

**Comment by maciejkula** [comment]
User ids are row indices into the features matrix: the features matrix you pass in has to have at least as many rows as you have users. You can have a look at the [docstring](https://github.com/lyst/lightfm/blob/master/lightfm/lightfm.py#L380).

**Comment by kyowill** [comment]
Yes, I know. My 'user_id' is i(user_id = 1, only one user).My 'feature_matrix' is 'user_features_concat[1]'(only one vector). But lightfm still reports the exception.In my opinion, it may be a bug:)

**Comment by maciejkula** [comment]
Python indexing is 0-based. By including a user id of 1, you are trying to index into the _second_ row of the features matrix. But, as you say, you have only one row.

**Comment by kyowill** [comment]
What I mean a user id of 1 is a index in some set, and  'user_features_concat[1]' is one of the row in features matrix which corresponding to the user id of 1. I didn't mean only one row.

**Comment by kyowill** [comment]
What I mean a user id of 1 is a index in some set, and 'user_features_concat[1]' is one of the row in features matrix which corresponding to the user id of 1. User id of 1 is the second user. feature matrix has many rows.

**Comment by maciejkula** [comment]
From your first example it looks like whatever you pass into predict has
only one row?

On 12 Jul 2017 08:50, "kyowill" <notifications@github.com> wrote:

What I mean a user id of 1 is a index in some set, and
'user_features_concat[1]' is one of the row in features matrix which
corresponding to the user id of 1. User id of 1 is the second user. feature
matrix has many rows.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/lyst/lightfm/issues/196#issuecomment-314684452>, or mute
the thread

**Comment by kyowill** [comment]
Yes, I only passed one row of features matrix. When user id of 0, I passed features matrix[0] in predict function, when user id of 1, I passed features matrix[1] in predict function. User id of 0 is OK, but user id of 1 is not. I couldn't find reason, so checked lightfm source code. when user id of 1, the 'n_user' is added up to 2, which bigger than number of feature vector('feature matrix[i]' is just one vector)
